### PR TITLE
Refactor recovery script

### PR DESCRIPTION
**Pull request by as2388**
- Modularises the different match strategies so it's easy to add, edit, and reorder the strategies that are applied.
 - Improves excel-mangled strategy to cover more cases.
 - Adds a new strategy for handling clipped messages

### CODE REVIEW

File: preprocess_recovered_hormuud_messages.py [hunk 1]
```diff
@@ -9,15 +11,71 @@
...
+
+        self.message_id = message_id
+        self.sender = sender
+        self.receiver = receiver
+        self.text = text
+        self.timestamp = timestamp
+
+    @classmethod
+    def from_hormuud_csv_row(cls, row):
+        try:
```
Comment (IsaackMwenda) [inline preprocess_recovered_hormuud_messages.py:51]: Nit:  add doc strings

File: preprocess_recovered_hormuud_messages.py [hunk 2]
```diff
@@ -44,48 +102,281 @@ def get_incoming_hormuud_messages_from_rapid_pro(google_cloud_credentials_file_p
...
+
+class Duplicates(MatchStrategy):
+    def __init__(self, csv_log_file_path=None):
+        """
+        Skips messages if they have the same text and sender as a message that has already been matched.
+        """
+        super().__init__("Duplicates", csv_log_file_path)
+
+    def skip_message(self, rapid_pro_message, matched_messages):
+        # A Rapid Pro message is a duplicate if contains the same text and sender as another Rapid Pro message
```
Comment (IsaackMwenda) [inline preprocess_recovered_hormuud_messages.py:261]:
```suggestion
        # A Rapid Pro message is a duplicate if contains the same text and sender has another Rapid Pro message
```

File: preprocess_recovered_hormuud_messages.py [hunk 3]
```diff
@@ -44,48 +102,281 @@ def get_incoming_hormuud_messages_from_rapid_pro(google_cloud_credentials_file_p
...
+            rapid_pro_text = rapid_pro_text.strip()  # numbers with whitespace -> just the number
+            if rapid_pro_text.startswith("0"):
+                rapid_pro_text = rapid_pro_text[1:]  # replace leading 0
+            if Decimal(rapid_pro_text) > 1000000000:
+                rapid_pro_text = f"{Decimal(rapid_pro_text):.4E}"  # big numbers -> scientific notation
+        if re.compile("^\".*\"$").match(rapid_pro_text):
+            rapid_pro_text = rapid_pro_text.replace("\"", "")  # strictly quoted text -> just the text
+        rapid_pro_text = rapid_pro_text.encode("ascii", "replace").decode("ascii")  # non-ascii characters -> '?'
+
+        recovered_text = recovered_message.text.encode("ascii", "replace").decode("ascii")  # non-ascii characters -> '?'
```
Comment (IsaackMwenda) [inline preprocess_recovered_hormuud_messages.py:241]: I think we also need to format the recovered_text  too here to match the formatted rapid_pro_text to cover messages that were actually sent as that e.g  startswith a 0?

**Review by IsaackMwenda — APPROVED** [review]
LGTM thanks.

**Review by DanielMwendwa — APPROVED** [review]

File: preprocess_recovered_hormuud_messages.py (hunk 1 above)
Comment (as2388) [reply preprocess_recovered_hormuud_messages.py:51]: added

File: preprocess_recovered_hormuud_messages.py (hunk 2 above)
Comment (as2388) [reply preprocess_recovered_hormuud_messages.py:261]: I did mean as here - but have rephrased to 'if they have the same text and the same sender as a message that has already been matched.' to be clearer

File: preprocess_recovered_hormuud_messages.py (hunk 3 above)
Comment (as2388) [reply preprocess_recovered_hormuud_messages.py:241]: From my experience looking at the last round of recovered data for WorldBank-SCD I don't think we do. If we received any messages that weren't affected by the excel-mangling described here (e.g. if we saw a message that did start with a 0), then it doesn't make sense to try to match it in the excel-mangling strategy. The other strategies should handle the case instead.

**Pull request merged**